internal/controller/supervisorconfig/githubupstreamwatcher/github_upstream_watcher.go

```diff
@@ -93,52 +100,210 @@ func (c *gitHubWatcherController) Sync(ctx controllerlib.Context) error {
+				gitHubClientSecretType,
+				clientIDDataKey,
+				clientSecretDataKey),
+		})
+	} else if applicationErr == nil {
+		clientID = string(secret.Data[clientIDDataKey])
+		clientSecret = string(secret.Data[clientSecretDataKey])
+		conditions = append(conditions, &metav1.Condition{
+			Type:    ClientCredentialsObtained,
+			Status:  metav1.ConditionTrue,
+			Reason:  upstreamwatchers.ReasonSuccess,
+			Message: "client credentials have been obtained, although they may not yet have been validated",
+		})
+	}
+
+	// Should not happen due to CRD defaulting
+	host := ""
+	if upstream.Spec.GitHubAPI.Host == nil || len(*upstream.Spec.GitHubAPI.Host) < 1 {
+		conditions = append(conditions, &metav1.Condition{
+			Type:    HostValid,
+			Status:  metav1.ConditionFalse,
+			Reason:  "InvalidHost",
+			Message: `spec.githubAPI.host ("") is not valid`,
+		})
+		conditions = append(conditions, &metav1.Condition{
+			Type:    GitHubConnectionValid,
+			Status:  metav1.ConditionUnknown,
+			Reason:  "UnableToValidate",
+			Message: `unable to determine if spec.githubAPI.host ("") is reachable or if TLS verification succeeds`,
+		})
+		conditions = append(conditions, &metav1.Condition{
+			Type:    TLSConfigurationValid,
+			Status:  metav1.ConditionUnknown,
+			Reason:  "UnableToValidate",
+			Message: `tls configuration cannot be validated for spec.githubAPI.host ("")`,
+		})
+	} else {
+		host = *upstream.Spec.GitHubAPI.Host
+		conditions = append(conditions, &metav1.Condition{
+			Type:    HostValid,
+			Status:  metav1.ConditionTrue,
+			Reason:  upstreamwatchers.ReasonSuccess,
+			Message: fmt.Sprintf("spec.githubAPI.host (%q) is valid", host),
+		})
+	}
+
+	// Should not happen due to CRD defaulting
+	if upstream.Spec.Claims.Groups == nil || upstream.Spec.Claims.Username == nil {
+		return nil, fmt.Errorf("spec.claims.groups and spec.claims.username are required")
+	}
+
+	// Should not happen due to CRD enum validation
+	switch *upstream.Spec.Claims.Username {
+	case v1alpha1.GitHubUsernameLoginAndID:
+	case v1alpha1.GitHubUsernameLogin:
+	case v1alpha1.GitHubUsernameID:
+		break
+	default:
+		return nil, fmt.Errorf("invalid spec.claims.username")
+	}
+
+	// Should not happen due to CRD enum validation
+	switch *upstream.Spec.Claims.Groups {
+	case v1alpha1.GitHubUseTeamNameForGroupName:
+	case v1alpha1.GitHubUseTeamSlugForGroupName:
+		break
+	default:
+		return nil, fmt.Errorf("invalid spec.claims.groups")
+	}
+
+	var policy v1alpha1.GitHubAllowedAuthOrganizationsPolicy
+	if upstream.Spec.AllowAuthentication.Organizations.Policy != nil {
+		policy = *upstream.Spec.AllowAuthentication.Organizations.Policy
+	}
+
+	// Should not happen due to CRD defaulting, enum validation, and CEL validation (for recent versions of K8s only!)
+	invalidPolicyCondition := &metav1.Condition{
+		Type:    OrganizationsPolicyValid,
+		Status:  metav1.ConditionFalse,
+		Reason:  "Invalid",
+		Message: `allowAuthentication policy for organizations is not valid`,
```

Comment:
Ideally message would indicate `why` it is not valid.

Reply:
updated